File: apps/mobile/services/uploader.ts
import NativeBackgroundLogger from '../modules/background-logger';
import { getPendingReadings, getJourneysByIds, markUploaded } from '../store/db';

// Set EXPO_PUBLIC_API_URL in .env for dev (e.g. http://localhost:8000).
// Defaults to production domain.
const API_BASE = process.env.EXPO_PUBLIC_API_URL ?? 'https://viesambcobertura.cat';


// Matches the actual SQLite column names returned by getPendingReadings
type DbReading = {
  id: string;
  journey_id: string | null;
  timestamp: number;
  lat_grid: number | null;
  lng_grid: number | null;
  signal_dbm: number | null;
  rsrp: number | null;
  rsrq: number | null;
  sinr: number | null;
  network_type: string | null;
  mcc: string | null;
  mnc: string | null;
  speed_kmh: number | null;
  gps_accuracy: number | null;
  pci: number | null;
  earfcn: number | null;
  band: number | null;
  cqi: number | null;
  timing_advance: number | null;
  cell_count: number | null;
  last_ping_ms: number | null;
};

async function isServerReachable(): Promise<boolean> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 5_000);
  const t0 = Date.now();
  try {
    const res = await fetch('https://clients3.google.com/generate_204', {
      method: 'HEAD',
      signal: controller.signal,
    });
    if (res.ok) {
      const latencyMs = Date.now() - t0;
      NativeBackgroundLogger.setLastPingMs(Date.now(), latencyMs).catch(() => {});
    }
    return res.ok;
  } catch {
    return false;
  } finally {
    clearTimeout(timer);
  }
}

const CHUNK_SIZE = 20;

function toPayloadReading(r: DbReading) {
  return {
    id: r.id,
    journey_id: r.journey_id ?? undefined,
    timestamp: new Date(r.timestamp).toISOString(),
    lat: r.lat_grid,
    lng: r.lng_grid,
    signal_dbm: r.signal_dbm,
    rsrp: r.rsrp,
    rsrq: r.rsrq,
    sinr: r.sinr,
    network_type: r.network_type,
    mcc: r.mcc,
    mnc: r.mnc,
    speed_kmh: r.speed_kmh,
    gps_accuracy: r.gps_accuracy,
    pci: r.pci,
    earfcn: r.earfcn,
    band: r.band,
    cqi: r.cqi,
    timing_advance: r.timing_advance,
    cell_count: r.cell_count,
    last_ping_ms: r.last_ping_ms,
    platform: 'android' as const,
    app_version: '0.1.0',
  };
}

export async function uploadPendingReadings(): Promise<{ uploaded: number; failed: number }> {
  if (!(await isServerReachable())) return { uploaded: 0, failed: 0 };

  const rows = (await getPendingReadings(500)) as unknown as DbReading[];
  if (rows.length === 0) return { uploaded: 0, failed: 0 };

  // Fetch journey metadata once for all rows (few journeys, many readings).
  const journeyIds = [...new Set(rows.map((r) => r.journey_id).filter(Boolean))] as string[];
  const journeyRows = await getJourneysByIds(journeyIds);
  const journeys = journeyRows.map((j) => ({
    id: j.id,
    started_at: new Date(j.started_at).toISOString(),
    ended_at: j.ended_at ? new Date(j.ended_at).toISOString() : null,
    platform: 'android' as const,
    app_version: '0.1.0',
  }));

  let uploaded = 0;
  let failed = 0;

  for (let i = 0; i < rows.length; i += CHUNK_SIZE) {
    const chunk = rows.slice(i, i + CHUNK_SIZE);
    try {
      const res = await fetch(`${API_BASE}/api/v1/readings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          journeys: i === 0 ? journeys : [],
          readings: chunk.map(toPayloadReading),
        }),
      });
      if (!res.ok) { failed += chunk.length; break; }
      const result = (await res.json()) as { accepted: number; rejected: number };
      await markUploaded(chunk.map((r) => r.id));
      uploaded += result.accepted;
      failed += result.rejected;
    } catch {
      failed += chunk.length;
      break;
    }
  }

  return { uploaded, failed };
}

File: apps/mobile/modules/background-logger/android/src/main/java/expo/modules/backgroundlogger/Uploader.kt
package expo.modules.backgroundlogger

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object Uploader {
    private const val TAG = "BackgroundLogger.Upload"
    private const val API_BASE = "https://viesambcobertura.cat"
    private const val CHUNK_SIZE = 500

    private val iso8601 = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    // Called from dbExecutor thread — runs synchronously.
    fun attemptUpload() {
        val rows = DbHelper.getPendingReadings(CHUNK_SIZE)
        if (rows.isEmpty()) return

        val ids = rows.mapNotNull { it["id"] as? String }
        val journeyIds = rows.mapNotNull { it["journey_id"] as? String }.distinct()
        val journeyRows = if (journeyIds.isNotEmpty()) DbHelper.getJourneysByIds(journeyIds) else emptyList()

        val payload = buildPayload(journeyRows, rows)

        try {
            val conn = URL("$API_BASE/api/v1/readings").openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json; charset=utf-8")
            conn.doOutput = true
            conn.connectTimeout = 10_000
            conn.readTimeout = 10_000
            conn.outputStream.use { it.write(payload.toString().toByteArray(Charsets.UTF_8)) }
            val status = conn.responseCode
            conn.disconnect()

            if (status in 200..299) {
                DbHelper.markUploaded(ids)
                Log.d(TAG, "Uploaded ${ids.size} readings (HTTP $status)")
            } else {
                Log.w(TAG, "Upload rejected HTTP $status")
            }
        } catch (e: Exception) {
            Log.d(TAG, "Upload failed: ${e.message}")
        }
    }

    private fun buildPayload(
        journeys: List<Map<String, Any?>>,
        readings: List<Map<String, Any?>>,
    ): JSONObject {
        val journeyArr = JSONArray()
        journeys.forEach { j ->
            journeyArr.put(JSONObject().apply {
                put("id", j["id"])
                put("started_at", msToIso(j["started_at"] as Long))
                val endedAt = j["ended_at"] as? Long
                put("ended_at", if (endedAt != null) msToIso(endedAt) else JSONObject.NULL)
                put("platform", "android")
                put("app_version", "0.1.0")
            })
        }

        val readingArr = JSONArray()
        readings.forEach { r ->
            readingArr.put(JSONObject().apply {
                put("id", r["id"])
                putNullable("journey_id", r["journey_id"])
                put("timestamp", msToIso(r["timestamp"] as Long))
                putNullable("lat", r["lat_grid"])
                putNullable("lng", r["lng_grid"])
                putNullable("signal_dbm", r["signal_dbm"])
                putNullable("rsrp", r["rsrp"])
                putNullable("rsrq", r["rsrq"])
                putNullable("sinr", r["sinr"])
                putNullable("network_type", r["network_type"])
                putNullable("mcc", r["mcc"])
                putNullable("mnc", r["mnc"])
                putNullable("speed_kmh", r["speed_kmh"])
                putNullable("gps_accuracy", r["gps_accuracy"])
                putNullable("pci", r["pci"])
                putNullable("earfcn", r["earfcn"])
                putNullable("band", r["band"])
                putNullable("cqi", r["cqi"])
                putNullable("timing_advance", r["timing_advance"])
                putNullable("cell_count", r["cell_count"])
                putNullable("last_ping_ms", r["last_ping_ms"])
                putNullable("last_ping_latency_ms", r["last_ping_latency_ms"])
                put("platform", "android")
                put("app_version", "0.1.0")
            })
        }

        return JSONObject().apply {
            put("journeys", journeyArr)
            put("readings", readingArr)
        }
    }

    private fun msToIso(ms: Long): String = iso8601.format(Date(ms))

    private fun JSONObject.putNullable(key: String, value: Any?) {
        if (value != null) put(key, value) else put(key, JSONObject.NULL)
    }
}
